[AIE2][AIE2P] Add combiners related to extract/insert/broadcast

andcarminati opened the pull request:
…x), index) into copy

Postlegalizer combiner that matches a pattern where:
 %18:_(<16 x s32>) = COPY $x0
 %10:_(<16 x s32>) = G_IMPLICIT_DEF
 %9:_(s32) = G_CONSTANT i32 0
 %8:_(s32) = G_AIE_SEXT_EXTRACT_VECTOR_ELT %18(<16 x s32>), %9(s32)
 %22:_(<16 x s32>) = G_AIE_INSERT_VECTOR_ELT %10, %8(s32), %9(s32)

And turns it into:
 %22:_(<16 x s32>) = COPY %18

FYI @martien-de-jong.

andcarminati committed:
[AIE2][AIE2P] Add combiner to convert insert(undef, extract(vec, index), index) into copy

Postlegalizer combiner that matches a pattern where:
 %18:_(<16 x s32>) = COPY $x0
 %10:_(<16 x s32>) = G_IMPLICIT_DEF
 %9:_(s32) = G_CONSTANT i32 0
 %8:_(s32) = G_AIE_SEXT_EXTRACT_VECTOR_ELT %18(<16 x s32>), %9(s32)
 %22:_(<16 x s32>) = G_AIE_INSERT_VECTOR_ELT %10, %8(s32), %9(s32)

And turns it into:
 %22:_(<16 x s32>) = COPY %18

martien-de-jong commented on llvm/lib/Target/AIE/AIECombine.td:
```diff
@@ -404,7 +411,8 @@ def AIE2PostLegalizerCustomCombiner
                                                           combine_extract_concat,
                                                           combine_unmerge_concat,
-                                                          combine_upd_to_concat, 
+                                                          combine_upd_to_concat,
```
whitespace change?

andcarminati replied:
Yeah, I removed one unnecessary space after...

F-Stuckmann replied:
have you guys figured out how to run clang format on td files?

andcarminati replied:
No, sorry! I just saw this space and deleted it....

martien-de-jong commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -4323,3 +4323,73 @@ bool llvm::matchMsbScalar(Register ScalarReg, Register BroadcastReg,
+
+  // Check that the insert source vector is G_IMPLICIT_DEF
+  const MachineInstr *InsertSrcMI = MRI.getVRegDef(InsertSrcVecReg);
+  if (!InsertSrcMI || InsertSrcMI->getOpcode() != TargetOpcode::G_IMPLICIT_DEF)
```
Oh, I have been using broadcast instead of insert into implicit def for the FMUL implementation. It will not work there.

andcarminati replied:
Humm, I was working on the FADD mir. It looks we use different approaches in different legalizations.

andcarminati committed:
[AIEX][GlobalISel] Add broadcast-extract-to-copy combiner for AIE2/AIE2P

This patch adds a new postlegalizer combiner that optimizes patterns where
an element is extracted from position 0, broadcast to a vector, and then
only position 0 is used through a chain of operations.

Pattern optimized:
  %extract = G_AIE_SEXT_EXTRACT_VECTOR_ELT %src(<N x T>), 0
  %broadcast = G_AIE_BROADCAST_VECTOR %extract
  ... (chain of concat/unmerge/vector ops)
  %result = G_AIE_SEXT_EXTRACT_VECTOR_ELT %final, 0

Transforms to:
  %extract = G_AIE_SEXT_EXTRACT_VECTOR_ELT %src(<N x T>), 0
  %broadcast = COPY %src  // Broadcast eliminated!
  ... (chain of operations)
  %result = G_AIE_SEXT_EXTRACT_VECTOR_ELT %final, 0

The combiner uses a conservative whitelist approach to verify that operations
in the chain don't shift vector elements. It handles:
- G_CONCAT_VECTORS (broadcast must be first operand)
- G_UNMERGE_VALUES (only first output used, others dead)
- Whitelisted operations: G_FADD, G_FMUL, G_FSUB, and specific AIE2P intrinsics
- Rejects: G_BITCAST, type mismatches, non-zero extracts, multiple uses

This optimization is particularly beneficial when broadcasts are used unnecessarily when
only the first element is needed.

The combiner runs as a postlegalizer custom combiner for both AIE2 and AIE2P.

andcarminati commented:
All commits from this PR were integrated as support in https://github.com/Xilinx/llvm-aie/pull/715.
